### ISSUE TITLE
Can not scan QRCode

### ISSUE THREAD

**Issue by dannyto** [metafloor/bwip-js#47]
Hi,

When i generate qrcode with this data :
"201601250031200000030042，戊二醛、苯扎溴铵溶液（水产用），兽药字（2013）170229276，金谷药业，07133387188"

It can not scan.

Thankz,

**Comment by metafloor** [comment]
I suspect this is either a utf-16/utf-8 conversion issue or a barcode code-page issue.  Please provide the url that you are using to generate the barcode.

**Comment by dannyto** [comment]
I just use your Online Barcode Generator page

![image](https://cloud.githubusercontent.com/assets/13778373/17995756/84d0404c-6b8e-11e6-98b9-47792879ab05.png)

And the output can not scan

**Comment by metafloor** [comment]
There is a bug in the code but even after fixing it, the generated barcode is still not scannable.

First lets fix the bug.  I will roll this into the next release, but for now, in the file bwipp.js, you need to add the following.  Look for the line around line number 25789:

```
        // Handle the `parse` option here rather than in BWIPP - eliminates
```

And add the following above it:

```
        // Convert utf-16 to utf-8 but leave binary (8-bit) strings untouched.
        if (/[\u0100-\uffff]/.test(text)) {
            text = unescape(encodeURIComponent(text));
        }

        // Handle the `parse` option here rather than in BWIPP - eliminates
```

I am at a loss why the resulting barcode is still unscannable.  For example, I tried the string

dragonfly 蜻蜓 1234

And it produced a valid barcode.  But your example string is still not scannable....

@terryburton:  Are you following this thread?

I tried the following in the BWIPP online demo and got the message **There was an error whilst processing the request**.   

The text string is from the initial issue report, converted to utf-8 and ^NNN encoded.  The options specified were `eclevel=M parse`.

201601250031200000030042^239^188^140^230^136^138^228^186^140^233^134^155^227^128^129^232^139^175^230^137^142^230^186^180^233^147^181^230^186^182^230^182^178^239^188^136^230^176^180^228^186^167^231^148^168^239^188^137^239^188^140^229^133^189^232^141^175^229^173^151^239^188^1362013^239^188^137170229276^239^188^140^233^135^145^232^176^183^232^141^175^228^184^154^239^188^14007133387188

Any insight you can provide?  Thanks!

**Comment by terryburton** [comment]
Confirmed that this is an upstream bug.

The encoder is incorrectly optimising the given data into the following mode sections:

```
/seq [
  0 [50 48 49 ... 52 50 239]       % ASCII mode: ordinal 239 doesn't belong here
  2 [188 140 230 ... 239 188 140]  % Byte mode
  0 [48 55 49 ... 49 56 56]        % ASCII
] def
```

I'll try to take a look this weekend.

**Comment by terryburton** [comment]
@metafloor Fixed by 6b9341b70d0b95b74259108d7854615866eed833 and released as 2016-08-27.

If you're interested and it helps to understand the somewhat opaque code, the optimisation is performed by an automata that operates a greedy character by character evaluation of a series of propositions (e.g. [X] "AbeforeN") which mean "there are at least X alphanumeric exclusive characters (A-Z except 0-9) before the next numeric exclusive character (0-9)" in order to determine the best mode for the upcoming data. The optimal numbers for each series of X values vary according to the symbol version that we are optimising for since things such as the character count indicator lengths vary between versions. The derivation for the optimal X values is here: https://github.com/bwipp/postscriptbarcode/blob/master/contrib/development/qrcode_proofs.txt

Since we don't know in advance what symbol size we will be selecting (we are aiming for the smallest symbol that fits the encoded data) we have to construct optimal sequences for each group of symbol versions that have differing properties, since different versions of the symbol may have different optimal encodings. Not pretty.

**Comment by terryburton** [comment]
I introduced a different bug patching QR Code which I've fixed in 2016-08-29.

**Comment by metafloor** [comment]
Released version 1.1.0 which contains the CLI from @alexeiskachykhin, the bug fix in bwip-js to convert utf-16 to utf-8, and BWIPP version 2016-08-29.

@terryburton Thank you for all your tireless work!

@dannyto Please give the online demo a try.  My testing indicates it is working correctly.

@alexeiskachykhin Thanks for the new CLI.

**Comment by dannyto** [comment]
It's working now

Thank you @terryburton , @metafloor ,@alexeiskachykhin.
